<a href="https://colab.research.google.com/github/arman-hossain45/Artificial_Neural_Network_Project/blob/main/NN_module_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# DL Assignment 03

**Name:Mohammad Arman hossain**

**Course Email:armanhossan890@gmail.com**  


## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Instruction video অনুযায়ী আমাদের দেয়া Colab ফাইলটি থেকে প্রথম একটি Save copy in drive করে নিবা। এরপর Google colab এর মধ্যে কোডগুলো করবে এবং সেই ফাইলটি ‘Anyone with the link’ & ‘View’ Access দিয়ে ফাইলটির Shareble Link টি সাবমিট করবে।

# General Instruction

You must choose your own dataset.

The dataset must:

Be a supervised learning dataset (Regression or Binary Classification)

Contain at least 300 samples

Have at least 2 input features

Be in CSV format

You are NOT allowed to use Dataset or DataLoader.

You must implement everything manually.

# Question 01: [ Marks 05 ]

## Dataset Preparation

## Using your chosen dataset:

Load the dataset.

Perform necessary preprocessing:

Handle missing values (if any)

Encode categorical variables (if necessary)

Feature scaling (if needed)

Separate features (X) and target (y).

Convert them into NumPy arrays.

Convert them into PyTorch tensors.

Split into training and testing sets.

Clearly explain each preprocessing decision.

# **Write** Answer 01:


In [ ]:
import numpy as np
import pandas as pd
import torch
import torch.nn as nn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder

In [ ]:
df =pd.read_csv("diabetics.csv")
df.head(10)

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1
5,5,116,74,0,0,25.6,0.201,30,0
6,3,78,50,32,88,31.0,0.248,26,1
7,10,115,0,0,0,35.3,0.134,29,0
8,2,197,70,45,543,30.5,0.158,53,1
9,8,125,96,0,0,0.0,0.232,54,1


In [ ]:
df.shape

(768, 9)

In [ ]:
df.isnull().sum()

,0
Pregnancies,0
Glucose,0
BloodPressure,0
SkinThickness,0
Insulin,0
BMI,0
DiabetesPedigreeFunction,0
Age,0
Outcome,0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [ ]:

df.describe().T

,count,mean,std,min,25%,50%,75%,max
Pregnancies,768.0,3.845052,3.369578,0.000,1.00000,3.0000,6.00000,17.00
Glucose,768.0,120.894531,31.972618,0.000,99.00000,117.0000,140.25000,199.00
BloodPressure,768.0,69.105469,19.355807,0.000,62.00000,72.0000,80.00000,122.00
SkinThickness,768.0,20.536458,15.952218,0.000,0.00000,23.0000,32.00000,99.00
Insulin,768.0,79.799479,115.244002,0.000,0.00000,30.5000,127.25000,846.00
BMI,768.0,31.992578,7.884160,0.000,27.30000,32.0000,36.60000,67.10
DiabetesPedigreeFunction,768.0,0.471876,0.331329,0.078,0.24375,0.3725,0.62625,2.42
Age,768.0,33.240885,11.760232,21.000,24.00000,29.0000,41.00000,81.00
Outcome,768.0,0.348958,0.476951,0.000,0.00000,0.0000,1.00000,1.00


In [ ]:

df.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [ ]:
x =df.drop("Outcome",axis=1)
y =df["Outcome"]

In [ ]:





x_train,x_test,y_train,y_test =train_test_split(x,y,test_size=0.2,random_state=42)

In [ ]:
scaler = StandardScaler()
x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

In [ ]:
x_train

array([[-0.52639686, -1.15139792, -3.75268255, ..., -4.13525578,
        -0.49073479, -1.03594038],
       [ 1.58804586, -0.27664283,  0.68034485, ..., -0.48916881,
         2.41502991,  1.48710085],
       [-0.82846011,  0.56687102, -1.2658623 , ..., -0.42452187,
         0.54916055, -0.94893896],
       ...,
       [ 1.8901091 , -0.62029661,  0.89659009, ...,  1.76054443,
         1.981245  ,  0.44308379],
       [-1.13052335,  0.62935353, -3.75268255, ...,  1.34680407,
        -0.78487662, -0.33992901],
       [-1.13052335,  0.12949347,  1.43720319, ..., -1.22614383,
        -0.61552223, -1.03594038]])

In [ ]:
y_train

,Outcome
60,0
618,1
346,0
294,0
231,1
...,...
71,0
106,0
270,1
435,1


In [ ]:
# numpy array to pytorch tensors
x_train_tensor= torch.from_numpy(x_train)
x_test_tensor= torch.from_numpy(x_test)
x_train_tensor= torch.tensor(x_train,dtype=torch.float32)
x_test_tensor= torch.tensor(x_test,dtype=torch.float32)

y_train_tensor= torch.tensor(y_train.values,dtype=torch.float32)
y_test_tensor= torch.tensor(y_test.values,dtype=torch.float32)


In [ ]:

x_train_tensor.shape

torch.Size([614, 8])

In [ ]:
y_train_tensor.shape

torch.Size([614])

# Question 02: [ Marks 20 ]

## Design a neural network using nn.Module.

### The model must contain:

Input layer

At least one hidden layer

Output layer

Suitable activation function



## Justify:

Number of hidden neurons

Choice of activation function

Print  the total number of trainable parameters.


In [ ]:
#Define  the  nn model
class MyNN(nn.Module):
  def __init__(self,num_features):
    super().__init__()
    self.model = nn.Sequential(
        nn.Linear(num_features,8),
        nn.ReLU(),
        nn.Linear(8,4),
        nn.ReLU(),
        nn.Linear(4,1),
        nn.Sigmoid()

  )

  def forward(self,features):
    return self.model(features)

model=MyNN(x_train_tensor.shape[1])

In [ ]:
x_train_tensor.shape

torch.Size([614, 8])

In [ ]:
!pip install torchinfo

In [ ]:
from torchinfo import summary
summary(model)

Layer (type:depth-idx)                   Param #
MyNN                                     --
├─Sequential: 1-1                        --
│    └─Linear: 2-1                       72
│    └─ReLU: 2-2                         --
│    └─Linear: 2-3                       36
│    └─ReLU: 2-4                         --
│    └─Linear: 2-5                       5
│    └─Sigmoid: 2-6                      --
Total params: 113
Trainable params: 113
Non-trainable params: 0

## Write Answer 02:

1 . Here  in code number of hidden layer is 2 first  hidden layer has 8 neuron and secound hidden layer has 4 neuron and 1 output which is predicted the disease

2.In this hidden layer  I use ReLU activation function and output layer I use sigmoid activation function

3.Here in this model we can see that totall trainable parameters is 113


# Question 03: [ Marks 10 ]

Choose an appropriate loss function.

Choose an optimizer.

<br>

Justify your choices based on:

Regression vs Classification

Nature of the dataset

In [ ]:
# import parameters
learning_rate =0.001
epochs =100

#define the  loss function

loss_function = nn.BCELoss()

#define the optimizer
optimizer = torch.optim.SGD(model.parameters(),lr=learning_rate)




## Write Answer 03:

1 . here I use the binary cross entropy loss function and also take learing rate is 0.001 and totall number of epochs is 100

2. This is a classification problem prediction of diabetics and also use BCE

3.I choose the SGD optimizer for the weight update

4. This data set is well define  there is no null values and its a classification problem set dataset for that reason  I use a
BCE for loss calculation

# Question 04: [ Marks 15 ]

## Implement a full training loop:

Forward pass

Loss computation

Backward pass

Parameter update

Gradient reset

### Requirements:

Train for at least 100 epochs.

Print loss every 10 epochs.

Store training loss history(You can pick your own Data Structure).

Explain clearly what happens in each step of the pipeline.

In [ ]:
#define the loop

for epoch in range(epochs):
  #forward pass
  y_pred = model(x_train_tensor)
  #Loss calculate
  loss = loss_function(y_pred,y_train_tensor.reshape(-1,1))

  #zero graadients

  optimizer.zero_grad()

  # backward pass

  loss.backward()

  # parameters upgrade

  optimizer.step()


  #print loss in each epoch
  if (epoch+1)%10==0:
     print(f"Epoch :{epoch},Loss :{loss.item()}")

Epoch :9,Loss :0.7223368883132935
Epoch :19,Loss :0.7218306660652161
Epoch :29,Loss :0.7213281989097595
Epoch :39,Loss :0.7208288908004761
Epoch :49,Loss :0.7203332185745239
Epoch :59,Loss :0.7198416590690613
Epoch :69,Loss :0.7193539142608643
Epoch :79,Loss :0.7188698053359985
Epoch :89,Loss :0.7183893322944641
Epoch :99,Loss :0.7179126143455505


## Write Answer 04:

* in this pipeline  I define a loop for every epochs(100) and insidde in this loop pass the forward pass and then calculate the loss then zero the gradients then pass the backward pass and upgrade the parametes and show the loss in every 10 epochs and at last show all the loss and use the SGD optimizer

# Question 05: [ Marks 10 ]

## Evaluate the model on test data.

## For regression:

Report MSE and MAE


## For classification:

Report Accuracy

Compare training vs testing performance.

State whether the model is underfitting or overfitting.

In [ ]:
with torch.no_grad():
  y_pred = model(x_train_tensor)
  y_pred = (y_pred>0.5).float()
  accuracy = (y_pred == y_test_tensor).float().mean()
  print(f"Accuracy: {accuracy.item()}")
  mse = ((y_pred - y_test_tensor)**2).mean()
  print(f"MSE: {mse.item()}")
  mae = (y_pred - y_test_tensor).abs().mean()
  print(f"MAE: {mae.item()}")


Accuracy: 0.3571428656578064
MSE: 0.6428571343421936
MAE: 0.6428571343421936


In [ ]:
# Now run this samae code for test data

#define the loop

for epoch in range(epochs):
  #forward pass
  y_pred2 = model(x_test_tensor)
  #Loss calculate
  loss = loss_function(y_pred2,y_test_tensor.reshape(-1,1))

  #zero graadients

  optimizer.zero_grad()

  # backward pass

  loss.backward()

  # parameters upgrade

  optimizer.step()


  #print loss in each epoch
  if (epoch+1)%10==0:
     print(f"Epoch :{epoch},Loss :{loss.item()}")

Epoch :9,Loss :0.7147321701049805
Epoch :19,Loss :0.7143145799636841
Epoch :29,Loss :0.7139002084732056
Epoch :39,Loss :0.7134886384010315
Epoch :49,Loss :0.7130804061889648
Epoch :59,Loss :0.7126749157905579
Epoch :69,Loss :0.7122723460197449
Epoch :79,Loss :0.7118725776672363
Epoch :89,Loss :0.7114737629890442
Epoch :99,Loss :0.7110778093338013


In [ ]:
with torch.no_grad():
  y_pred2 = model(x_test_tensor)
  y_pred2 = (y_pred>0.5).float()
  accuracy2 = (y_pred2 == y_test_tensor).float().mean()
  print(f"Accuracy: {accuracy2.item()}")



Accuracy: 0.3571428656578064


We see that our model is not  over fit or under fit because show samee accuracy on test data and train data set

# Question 06: [ Marks 20 ]

## Modify at least ONE of the following:

Learning rate

Number of hidden neurons

Number of epochs

### Train again and compare:

Convergence speed

Final performance

Explain how the change affected the model.

# Her is below given the modification is here

1. epochs = 200

2. Learninf RAte = 0.1

3.increase the layer neuron and also increase the hidden layer also input a hidden layer in this existing system here is the code is

In [ ]:
#Define  the  nn model
class MyNN(nn.Module):
  def __init__(self,num_features):
    super().__init__()
    self.model = nn.Sequential(
        nn.Linear(num_features,16),
        nn.ReLU(),
        nn.Linear(16,8),
        nn.ReLU(),
        nn.Linear(8,4),
        nn.ReLU(),
        nn.Linear(4,1),
        nn.Sigmoid()

  )

  def forward(self,features):
    return self.model(features)

model=MyNN(x_train_tensor.shape[1])

In [ ]:
# import parameters
learning_rate2 =0.1
epochs2 =200

#define the  loss function

loss_function = nn.BCELoss()

#define the optimizer
optimizer = torch.optim.SGD(model.parameters(),lr=learning_rate2)





for epoch in range(epochs2):
  #forward pass
  y_pred = model(x_train_tensor)
  #Loss calculate
  loss = loss_function(y_pred,y_train_tensor.reshape(-1,1))

  #zero graadients

  optimizer.zero_grad()

  # backward pass

  loss.backward()

  # parameters upgrade

  optimizer.step()


  #print loss in each epoch
  if (epoch+1)%10==0:
     print(f"Epoch :{epoch},Loss :{loss.item()}")

Epoch :9,Loss :0.6756730675697327
Epoch :19,Loss :0.6598787307739258
Epoch :29,Loss :0.6514642238616943
Epoch :39,Loss :0.6465875506401062
Epoch :49,Loss :0.6433086395263672
Epoch :59,Loss :0.6407585740089417
Epoch :69,Loss :0.6384431719779968
Epoch :79,Loss :0.636128306388855
Epoch :89,Loss :0.6336124539375305
Epoch :99,Loss :0.6307276487350464
Epoch :109,Loss :0.6272875070571899
Epoch :119,Loss :0.623085618019104
Epoch :129,Loss :0.6178396344184875
Epoch :139,Loss :0.6112114787101746
Epoch :149,Loss :0.6030386090278625
Epoch :159,Loss :0.592991292476654
Epoch :169,Loss :0.5811353325843811
Epoch :179,Loss :0.567524254322052
Epoch :189,Loss :0.5524306893348694
Epoch :199,Loss :0.5367754101753235


In [ ]:
with torch.no_grad():
  y_pred = model(x_train_tensor)
  y_pred = (y_pred>0.5).float()
  accuracy = (y_pred == y_test_tensor).float().mean()
  print(f"Accuracy: {accuracy.item()}")
  mse = ((y_pred - y_test_tensor)**2).mean()
  print(f"MSE: {mse.item()}")
  mae = (y_pred - y_test_tensor).abs().mean()
  print(f"MAE: {mae.item()}")


Accuracy: 0.5967891812324524
MSE: 0.4032107889652252
MAE: 0.4032107889652252


## Write Answer 06:

1.Before we see that the model accuracy is 0.35 percent where we use only 100 epochs and learning rate is 0.001 and use only 2 hidden layer so our model perform dont well


2.In this term our model is upgrade wher we use 200 epochs and also learning rate is 0.1 and also use extra hidden layer with this existing layer so the model  perform well which is 0.59 our model don well before performance  


So we say that increase the hidden layer and increase the epochs model is perform well we say that


# Question 07: [ Marks 20 ]


# Training Analysis

Answer the following:

Why must gradients be reset every epoch?

What happens if learning rate is too high?

What happens if learning rate is too small?

Why do we define layers inside the constructor (__init__) and not inside forward()?


## Write Answer 07:

#Answer to the qus no 1:

WE know that in pytorch gradients are accumulated by default ,when we call loss.backward() the computed gradients are added to the existing gradients store in the parameters  

If we dont reset the gradients before the next epoch or iteration the new gradients will be added to the previous ones. This causes incorrect weight updates and can lead to unstable trainig

There fore gradients must be reset


#Answer to the question no 2

What happen if the learning rate is too high

if the learning rate is too high model upgrades its wei9ght with very large steps during training

1.training can be unstable

2.The optimizer may overshoot the optimal solution

3.THe loss may diverge or explode


#Answer to the question no 3

what happen if the learning rate is too small


If the learning rate is too small , the model upgrades its weights with very small steps

so this problem is occur

1.Very slow training

2.The model taking a long time to converge

3.The possibility of getting stuck in local minima


Training process becomes inefficient and also time consuming


#Answer to the question no 4

Layers are defined inside the constructor  __init__ so  that they are created only once when the model is initialized

This allows pytorch to

1.Track weights and biases

2.Upgrade parameters during training


If layer  were defined inside the forward() function

1.A new layer would be reinitialized repeatedly

2.the model would not learn properly